### PR TITLE
NoMethodError: undefined method `user' for nil:NilClass

### LINKED ISSUE #55: NoMethodError: undefined method `request_uri' for #<URI::Generic >
View details in Rollbar: [https://rollbar.com/yaroslavtupitskyi/ProjectR/items/15/](https://rollbar.com/yaroslavtupitskyi/ProjectR/items/15/)

```

NoMethodError: undefined method `request_uri' for #<URI::Generic >
  File "/home/deploy/releases/20161125133846/app/jobs/sender_job.rb", line 8, in perform
  File "/home/deploy/releases/20161125133846/app/controllers/projects/test_messages_controller.rb", line 3, in create
  File "/home/deploy/shared/bundle/ruby/2.3.0/gems/actionpack-5.0.0/lib/action_controller/metal/basic_implicit_render.rb", line 4, in send_action
  File "/home/deploy/shared/bundle/ruby/2.3.0/gems/actionpack-5.0.0/lib/abstract_controller/base.rb", line 188, in process_action
  File "/home/deploy/shared/bundle/ruby/2.3.0/gems/actionpack-5.0.0/lib/action_controller/metal/rendering.rb", line 30, in process_action
  File "/home/deploy/shared/bundle/ruby/2.3.0/gems/actionpack-5.0.0/lib/abstract_controller/callbacks.rb", line 20, in block in process_action
  File "/home/deploy/shared/bundle/ruby/2.3.0/gems/activesupport-5.0.0/lib/active_support/callbacks.rb", line 126, in call
  File "/home/deploy/shared/bundle/ruby/2.3.0/gems/activesupport-5.0.0/lib/active_support/callbacks.rb", line 506, in block (2 levels) in compile
  File "/home/deploy/shared/bundle/ruby/2.3.0/gems/activesupport-5.0.0/lib/active_support/callbacks.rb", line 455, in call
  File "/home/deploy/shared/bundle/ruby/2.3.0/gems/activesupport-5.0.0/lib/active_support/callbacks.rb", line 101, in __run_callbacks__
  File "/home/deploy/shared/bundle/ruby/2.3.0/gems/activesupport-5.0.0/lib/active_support/callbacks.rb", line 750, in _run_process_action_callbacks
  File "/home/deploy/shared/bundle/ruby/2.3.0/gems/activesupport-5.0.0/lib/active_support/callbacks.rb", line 90, in run_callbacks
  File "/home/deploy/shared/bundle/ruby/2.3.0/gems/actionpack-5.0.0/lib/abstract_controller/callbacks.rb", line 19, in process_action
  File "/home/deploy/shared/bundle/ruby/2.3.0/gems/actionpack-5.0.0/lib/action_controller/metal/rescue.rb", line 20, in process_action
  File "/home/deploy/shared/bundle/ruby/2.3.0/gems/actionpack-5.0.0/lib/action_controller/metal/instrumentation.rb", line 32, in block in process_action
  File "/home/deploy/shared/bundle/ruby/2.3.0/gems/activesupport-5.0.0/lib/active_support/notifications.rb", line 164, in block in instrument
  File "/home/deploy/shared/bundle/ruby/2.3.0/gems/activesupport-5.0.0/lib/active_support/notifications/instrumenter.rb", line 21, in instrument
  File "/home/deploy/shared/bundle/ruby/2.3.0/gems/activesupport-5.0.0/lib/active_support/notifications.rb", line 164, in instrument
  File "/home/deploy/shared/bundle/ruby/2.3.0/gems/actionpack-5.0.0/lib/action_controller/metal/instrumentation.rb", line 30, in process_action
  File "/home/deploy/shared/bundle/ruby/2.3.0/gems/actionpack-5.0.0/lib/action_controller/metal/params_wrapper.rb", line 248, in process_action
  File "/home/deploy/shared/bundle/ruby/2.3.0/gems/activerecord-5.0.0/lib/active_record/railties/controller_runtime.rb", line 18, in process_action
  File "/home/deploy/shared/bundle/ruby/2.3.0/gems/actionpack-5.0.0/lib/abstract_controller/base.rb", line 126, in process
  File "/home/deploy/shared/bundle/ruby/2.3.0/gems/actionview-5.0.0/lib/action_view/rendering.rb", line 30, in process
  File "/home/deploy/shared/bundle/ruby/2.3.0/gems/actionpack-5.0.0/lib/action_controller/metal.rb", line 190, in dispatch
  File "/home/deploy/shared/bundle/ruby/2.3.0/gems/actionpack-5.0.0/lib/action_controller/metal.rb", line 262, in dispatch
  File "/home/deploy/shared/bundle/ruby/2.3.0/gems/actionpack-5.0.0/lib/action_dispatch/routing/route_set.rb", line 50, in dispatch
  File "/home/deploy/shared/bundle/ruby/2.3.0/gems/actionpack-5.0.0/lib/action_dispatch/routing/route_set.rb", line 32, in serve
  File "/home/deploy/shared/bundle/ruby/2.3.0/gems/actionpack-5.0.0/lib/action_dispatch/journey/router.rb", line 39, in block in serve
  File "/home/deploy/shared/bundle/ruby/2.3.0/gems/actionpack-5.0.0/lib/action_dispatch/journey/router.rb", line 26, in each
  File "/home/deploy/shared/bundle/ruby/2.3.0/gems/actionpack-5.0.0/lib/action_dispatch/journey/router.rb", line 26, in serve
  File "/home/deploy/shared/bundle/ruby/2.3.0/gems/actionpack-5.0.0/lib/action_dispatch/routing/route_set.rb", line 725, in call
  File "/home/deploy/shared/bundle/ruby/2.3.0/gems/rack-2.0.1/lib/rack/etag.rb", line 25, in call
  File "/home/deploy/shared/bundle/ruby/2.3.0/gems/rack-2.0.1/lib/rack/conditional_get.rb", line 38, in call
  File "/home/deploy/shared/bundle/ruby/2.3.0/gems/rack-2.0.1/lib/rack/head.rb", line 12, in call
  File "/home/deploy/shared/bundle/ruby/2.3.0/gems/rack-2.0.1/lib/rack/session/abstract/id.rb", line 222, in context
  File "/home/deploy/shared/bundle/ruby/2.3.0/gems/rack-2.0.1/lib/rack/session/abstract/id.rb", line 216, in call
  File "/home/deploy/shared/bundle/ruby/2.3.0/gems/actionpack-5.0.0/lib/action_dispatch/middleware/cookies.rb", line 613, in call
  File "/home/deploy/shared/bundle/ruby/2.3.0/gems/actionpack-5.0.0/lib/action_dispatch/middleware/callbacks.rb", line 38, in block in call
  File "/home/deploy/shared/bundle/ruby/2.3.0/gems/activesupport-5.0.0/lib/active_support/callbacks.rb", line 97, in __run_callbacks__
  File "/home/deploy/shared/bundle/ruby/2.3.0/gems/activesupport-5.0.0/lib/active_support/callbacks.rb", line 750, in _run_call_callbacks
  File "/home/deploy/shared/bundle/ruby/2.3.0/gems/activesupport-5.0.0/lib/active_support/callbacks.rb", line 90, in run_callbacks
  File "/home/deploy/shared/bundle/ruby/2.3.0/gems/actionpack-5.0.0/lib/action_dispatch/middleware/callbacks.rb", line 36, in call
  File "/home/deploy/shared/bundle/ruby/2.3.0/gems/actionpack-5.0.0/lib/action_dispatch/middleware/remote_ip.rb", line 79, in call
  File "/home/deploy/shared/bundle/ruby/2.3.0/gems/rollbar-2.12.0/lib/rollbar/middleware/rails/rollbar.rb", line 24, in block in call
  File "/home/deploy/shared/bundle/ruby/2.3.0/gems/rollbar-2.12.0/lib/rollbar.rb", line 735, in scoped
  File "/home/deploy/shared/bundle/ruby/2.3.0/gems/rollbar-2.12.0/lib/rollbar/middleware/rails/rollbar.rb", line 22, in call
  File "/home/deploy/shared/bundle/ruby/2.3.0/gems/actionpack-5.0.0/lib/action_dispatch/middleware/debug_exceptions.rb", line 49, in call
  File "/home/deploy/shared/bundle/ruby/2.3.0/gems/rollbar-2.12.0/lib/rollbar/middleware/rails/show_exceptions.rb", line 22, in call_with_rollbar
  File "/home/deploy/shared/bundle/ruby/2.3.0/gems/actionpack-5.0.0/lib/action_dispatch/middleware/show_exceptions.rb", line 31, in call
  File "/home/deploy/shared/bundle/ruby/2.3.0/gems/railties-5.0.0/lib/rails/rack/logger.rb", line 36, in call_app
  File "/home/deploy/shared/bundle/ruby/2.3.0/gems/railties-5.0.0/lib/rails/rack/logger.rb", line 24, in block in call
  File "/home/deploy/shared/bundle/ruby/2.3.0/gems/activesupport-5.0.0/lib/active_support/tagged_logging.rb", line 70, in block in tagged
  File "/home/deploy/shared/bundle/ruby/2.3.0/

### LINKED ISSUE #64: NoMethodError: undefined method `user' for nil:NilClass
View details in Rollbar: [https://rollbar.com/yaroslavtupitskyi/ProjectR/items/20/](https://rollbar.com/yaroslavtupitskyi/ProjectR/items/20/)

```

NoMethodError: undefined method `user' for nil:NilClass
  File "/home/deploy/releases/20170105125254/app/jobs/application_job.rb", line 11, in status
  File "/home/deploy/releases/20170105125254/app/jobs/sender_job.rb", line 27, in attachments
  File "/home/deploy/releases/20170105125254/app/jobs/sender_job.rb", line 19, in params_message
  File "/home/deploy/releases/20170105125254/app/jobs/sender_job.rb", line 9, in perform
  File "/home/deploy/releases/20170105125254/app/jobs/rebaser_job.rb", line 28, in rebase_pr
  File "/home/deploy/releases/20170105125254/app/jobs/rebaser_job.rb", line 10, in perform
  File "/home/deploy/releases/20170105125254/app/controllers/rebases_controller.rb", line 34, in pull_request
  File "/home/deploy/releases/20170105125254/app/controllers/rebases_controller.rb", line 11, in create
  File "/home/deploy/shared/bundle/ruby/2.3.0/gems/actionpack-5.0.0/lib/action_controller/metal/basic_implicit_render.rb", line 4, in send_action
  File "/home/deploy/shared/bundle/ruby/2.3.0/gems/actionpack-5.0.0/lib/abstract_controller/base.rb", line 188, in process_action
  File "/home/deploy/shared/bundle/ruby/2.3.0/gems/actionpack-5.0.0/lib/action_controller/metal/rendering.rb", line 30, in process_action
  File "/home/deploy/shared/bundle/ruby/2.3.0/gems/actionpack-5.0.0/lib/abstract_controller/callbacks.rb", line 20, in block in process_action
  File "/home/deploy/shared/bundle/ruby/2.3.0/gems/activesupport-5.0.0/lib/active_support/callbacks.rb", line 126, in call
  File "/home/deploy/shared/bundle/ruby/2.3.0/gems/activesupport-5.0.0/lib/active_support/callbacks.rb", line 506, in block (2 levels) in compile
  File "/home/deploy/shared/bundle/ruby/2.3.0/gems/activesupport-5.0.0/lib/active_support/callbacks.rb", line 455, in call
  File "/home/deploy/shared/bundle/ruby/2.3.0/gems/activesupport-5.0.0/lib/active_support/callbacks.rb", line 101, in __run_callbacks__
  File "/home/deploy/shared/bundle/ruby/2.3.0/gems/activesupport-5.0.0/lib/active_support/callbacks.rb", line 750, in _run_process_action_callbacks
  File "/home/deploy/shared/bundle/ruby/2.3.0/gems/activesupport-5.0.0/lib/active_support/callbacks.rb", line 90, in run_callbacks
  File "/home/deploy/shared/bundle/ruby/2.3.0/gems/actionpack-5.0.0/lib/abstract_controller/callbacks.rb", line 19, in process_action
  File "/home/deploy/shared/bundle/ruby/2.3.0/gems/actionpack-5.0.0/lib/action_controller/metal/rescue.rb", line 20, in process_action
  File "/home/deploy/shared/bundle/ruby/2.3.0/gems/actionpack-5.0.0/lib/action_controller/metal/instrumentation.rb", line 32, in block in process_action
  File "/home/deploy/shared/bundle/ruby/2.3.0/gems/activesupport-5.0.0/lib/active_support/notifications.rb", line 164, in block in instrument
  File "/home/deploy/shared/bundle/ruby/2.3.0/gems/activesupport-5.0.0/lib/active_support/notifications/instrumenter.rb", line 21, in instrument
  File "/home/deploy/shared/bundle/ruby/2.3.0/gems/activesupport-5.0.0/lib/active_support/notifications.rb", line 164, in instrument
  File "/home/deploy/shared/bundle/ruby/2.3.0/gems/actionpack-5.0.0/lib/action_controller/metal/instrumentation.rb", line 30, in process_action
  File "/home/deploy/shared/bundle/ruby/2.3.0/gems/actionpack-5.0.0/lib/action_controller/metal/params_wrapper.rb", line 248, in process_action
  File "/home/deploy/shared/bundle/ruby/2.3.0/gems/activerecord-5.0.0/lib/active_record/railties/controller_runtime.rb", line 18, in process_action
  File "/home/deploy/shared/bundle/ruby/2.3.0/gems/actionpack-5.0.0/lib/abstract_controller/base.rb", line 126, in process
  File "/home/deploy/shared/bundle/ruby/2.3.0/gems/actionview-5.0.0/lib/action_view/rendering.rb", line 30, in process
  File "/home/deploy/shared/bundle/ruby/2.3.0/gems/actionpack-5.0.0/lib/action_controller/metal.rb", line 190, in dispatch
  File "/home/deploy/shared/bundle/ruby/2.3.0/gems/actionpack-5.0.0/lib/action_controller/metal.rb", line 262, in dispatch
  File "/home/deploy/shared/bundle/ruby/2.3.0/gems/actionpack-5.0.0/lib/action_dispatch/routing/route_set.rb", line 50, in dispatch
  File "/home/deploy/shared/bundle/ruby/2.3.0/gems/actionpack-5.0.0/lib/action_dispatch/routing/route_set.rb", line 32, in serve
  File "/home/deploy/shared/bundle/ruby/2.3.0/gems/actionpack-5.0.0/lib/action_dispatch/journey/router.rb", line 39, in block in serve
  File "/home/deploy/shared/bundle/ruby/2.3.0/gems/actionpack-5.0.0/lib/action_dispatch/journey/router.rb", line 26, in each
  File "/home/deploy/shared/bundle/ruby/2.3.0/gems/actionpack-5.0.0/lib/action_dispatch/journey/router.rb", line 26, in serve
  File "/home/deploy/shared/bundle/ruby/2.3.0/gems/actionpack-5.0.0/lib/action_dispatch/routing/route_set.rb", line 725, in call
  File "/home/deploy/shared/bundle/ruby/2.3.0/gems/rack-2.0.1/lib/rack/etag.rb", line 25, in call
  File "/home/deploy/shared/bundle/ruby/2.3.0/gems/rack-2.0.1/lib/rack/conditional_get.rb", line 38, in call
  File "/home/deploy/shared/bundle/ruby/2.3.0/gems/rack-2.0.1/lib/rack/head.rb", line 12, in call
  File "/home/deploy/shared/bundle/ruby/2.3.0/gems/rack-2.0.1/lib/rack/session/abstract/id.rb", line 222, in context
  File "/home/deploy/shared/bundle/ruby/2.3.0/gems/rack-2.0.1/lib/rack/session/abstract/id.rb", line 216, in call
  File "/home/deploy/shared/bundle/ruby/2.3.0/gems/actionpack-5.0.0/lib/action_dispatch/middleware/cookies.rb", line 613, in call
  File "/home/deploy/shared/bundle/ruby/2.3.0/gems/actionpack-5.0.0/lib/action_dispatch/middleware/callbacks.rb", line 38, in block in call
  File "/home/deploy/shared/bundle/ruby/2.3.0/gems/activesupport-5.0.0/lib/active_support/callbacks.rb", line 97, in __run_callbacks__
  File "/home/deploy/shared/bundle/ruby/2.3.0/gems/activesupport-5.0.0/lib/active_support/callbacks.rb", line 750, in _run_call_callbacks
  File "/home/deploy/shared/bundle/ruby/2.3.0/gems/activesupport-5.0.0/lib/active_support/callbacks.rb", line 90, in run_callbacks
  File "/home/deploy/shared/bundle/ruby/2.3.0/gems/actionpack-5.0.0/lib/action_dispatch/middleware/callbacks.rb", line 36, in call
  File "/home/deploy/shared/bundle/ruby/2.3.0/gems/actionpack-5.0.0/lib/action_dispatch/middleware/remote_ip.rb", line 79, in call
  File "/home/deploy/shared/bundle/ruby/2.3.0/gems/rollbar-2.12.0/lib/rollbar/middleware/rails/rollbar.rb", line 24, in block in call
  File "/home/deploy/shared/bundle/ruby/2.3.0/gems/rollbar-2.12.0/lib/rollbar.rb", line 735, in scoped
  File "/home/deploy/shared/bundle/ruby/2.3.0/gems/rollbar-2.12.0/lib/rollbar/middleware/rails/rollbar.rb", line 22, in call
  File "/home/deploy/shared/bundle/ruby/2.3.0/gems/actionpack-5.0.0/lib/action_dispatch/middleware/debug_exceptions.rb", line 49, in call
  File "/home/deploy/shared/bundle/ruby/2.3.0/gems/rollbar-2.12.0/lib/rollbar/middleware/rails/show_exceptions.rb", line 22, in call_with_rollbar

### PR DESCRIPTION
**Pull request by Sky7hunder**
Close #64, close #55

### CODE REVIEW

File: app/jobs/sender_job.rb [hunk 1]
```diff
@@ -2,6 +2,8 @@ class SenderJob < ApplicationJob
   queue_as :default
 
   def perform(attr)
+    @rebase = attr[:rebase]
+    @repo = attr[:repo]
```
Comment (galulex) [inline app/jobs/sender_job.rb]: why do we need this?

File: spec/jobs/sender_job_spec.rb [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+      create_status: status,
+      status: status,
+      pulls: [pull_request],
+      pull: pull_request
+    )
+  end
+
+  before { allow(Octokit::Client).to receive(:new).and_return(github) }
+
+  context 'when was sanded' do
```
Comment (galulex) [inline spec/jobs/sender_job_spec.rb]: http://betterspecs.org

Comment (galulex) [inline spec/jobs/sender_job_spec.rb]: I am not sure that this example really check something. It looks like you check your own double

File: app/jobs/sender_job.rb (hunk 1 above)
Comment (Sky7hunder) [inline app/jobs/sender_job.rb]: I detect that previous instance variables are not available in `SenderJob`. When we call method `status` that uses this variables it fails

File: spec/jobs/sender_job_spec.rb (hunk 2 above)
Comment (Sky7hunder) [inline spec/jobs/sender_job_spec.rb]: I have tried to check success work of `SenderJob`

File: app/jobs/sender_job.rb (hunk 1 above)
Comment (galulex) [inline app/jobs/sender_job.rb]: Link to code where we use `@repo` please

Comment (Sky7hunder) [inline app/jobs/sender_job.rb]: yep, i see. It looks like used nowhere

File: spec/jobs/sender_job_spec.rb [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+      pull: pull_request
+    )
+  end
+  let(:channel_url) { 'https://hooks.slack.com' }
+  let(:options) { { repo: repo, rebase: rebase, status: 'success' } }
+
+  before { allow(Octokit::Client).to receive(:new).and_return(github) }
+
+  context 'when channel url successful' do
+    it { expect(described_class.perform_now(options)).to be_instance_of(Net::HTTPFound) }
```
Comment (galulex) [inline spec/jobs/sender_job_spec.rb]: it `has valid status`

File: app/jobs/sender_job.rb [hunk 4]
```diff
@@ -2,12 +2,15 @@ class SenderJob < ApplicationJob
...
 
   def perform(attr)
+    @rebase = attr[:rebase]
     uri = URI.parse(attr[:repo][:channel_url])
-    http = Net::HTTP.new(uri.host, uri.port)
-    http.use_ssl = true
-    request = Net::HTTP::Post.new(uri.request_uri)
-    request.body = params_message(attr).to_json
-    http.request(request)
+    if uri.respond_to?(:request_uri)
```
Comment (galulex) [inline app/jobs/sender_job.rb]: return unless `uri.respond_to?(:request_uri)`

**Pull request merged**